### PR TITLE
Skip lambda cached field

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Fixes https://github.com/tonerdo/coverlet/issues/730

The issue is related to lambda cached fields, that triggers another issue on coverage calculation.
I've also refactored a bit `GetBranchPoints` to keep code flow clear.

cc: @matteoerigozzi can you take a look?

### CODE REVIEW

File: test/coverlet.core.tests/Coverage/CoverageTests.AsyncAwait.cs [hunk 1]
```diff
@@ -33,33 +33,32 @@ public void AsyncAwait()
...
+                                    (74, 0), (75, 0), (76, 0), (77, 0), (78, 0),
+                                    // ContinuationCalled -> line 83 should be 1 hit some issue with Continuation state machine
+                                    (81, 1), (82, 1), (83, 2), (84, 1), (85, 1),
+                                    // ConfigureAwait
+                                    (89, 1), (90, 1)
+                                    )
+                .AssertBranchesCovered(BuildConfiguration.Debug, (16, 0, 0), (16, 1, 1), (43, 0, 3), (43, 1, 1), (43, 2, 1), (43, 3, 1), (43, 4, 0))
+                // Real branch should be 2, we should try to remove compiler generated branch in method ContinuationNotCalled/ContinuationCalled
+                // for Continuation state machine
+                .ExpectedTotalNumberOfBranches(BuildConfiguration.Debug, 2);
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.AsyncAwait.cs]: fixed also this bug

File: test/coverlet.core.tests/Coverage/CoverageTests.Lambda.cs [hunk 2]
```diff
@@ -27,23 +27,45 @@ public void Lambda_Issue343()
...
+                    {
+                        ((Task)instance.Invoke()).ConfigureAwait(false).GetAwaiter().GetResult();
+                        return Task.CompletedTask;
+                    },
+                    persistPrepareResultToFile: pathSerialize, disableRestoreModules: true);
+
+                    return 0;
+                }, path, invokeInProcess: true).Dispose();
 
-                        // Unexpected branches - generated by compiler to cache delegate instance
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.Lambda.cs]: fixed also this bug

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 3]
```diff
@@ -92,17 +98,148 @@ private static bool IsRecognizedMoveNextInsideAsyncStateMachineProlog(MethodDefi
...
+                        operand.DeclaringType.Scope.Name == "System.Threading.Tasks.Extensions"
+                    )
+               )
+            {
+                return true;
+            }
+            return false;
+        }
+
+        private static bool SkipLambdaCachedField(Instruction instruction)
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Key method

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 4]
```diff
@@ -36,6 +36,11 @@ private static bool IsCompilerGenerated(MethodDefinition methodDefinition)
             return false;
         }
 
+        private static bool IsCompilerGenerated(FieldDefinition fieldDefinition)
+        {
+            return fieldDefinition.DeclaringType.CustomAttributes.Any(ca => ca.AttributeType.FullName == typeof(CompilerGeneratedAttribute).FullName);
```
Comment (matteoerigozzi) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Could be inside nested class? In this case it's more correct to search up to the outermost declaring type as done for ```IsCompilerGenerated(MethodDefinition methodDefinition)``` overload.

Comment (MarcoRossignoli) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: I'm not sure if it's possible...do you think it's possible?

Comment (MarcoRossignoli) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: I'll merge as is for now I'm not sure if a field for a lambda could be nested inside other generated type...I cannot find samples, if you find it let me know and we'll amend.

**Pull request merged**